File: androidApp/src/main/java/com/wonddak/loacell/android/viewModel/LoaCellViewModel.kt
package com.wonddak.loacell.android.viewModel


import androidx.compose.material3.SnackbarDuration
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.viewModelScope
import com.wonddak.loacell.database.AppDataBase
import com.wonddak.loacell.CommonViewModel
import com.wonddak.loacell.Config
import com.wonddak.loacell.ILOA
import com.wonddak.loacell.RoomInfo
import com.wonddak.loacell.auth.FBUser
import com.wonddak.loacell.auth.LoginHelper
import com.wonddak.loacell.ext.TotalRoomInfo
import com.wonddak.loacell.storage.SynergyReferenceHelper
import kotlinx.coroutines.launch

class LoaCellViewModel(
    dataBase: AppDataBase,
    config: Config,
    loginHelper: LoginHelper,
    synergyReferenceHelper: SynergyReferenceHelper
) : CommonViewModel(dataBase, config, loginHelper,synergyReferenceHelper) {

    //region snackbar
    private val snackBarController = SnackBarController()

    val snackBarMessage
        get() = snackBarController.snackBarMessage

    fun showSnackBar(
        message: String,
        label: String? = null,
        duration: SnackbarDuration = SnackbarDuration.Short,
        action: () -> Unit = { resetSnackBar() },
    ) = snackBarController.showSnackBar(message, label, duration, action)

    override fun showSnackBar(msg: String) {
        showSnackBar(msg, label = "확인")
    }

    fun resetSnackBar() = snackBarController.resetSnackBar()
    //endregion

    var showSetting by mutableStateOf(false)
    override fun getSetting(): Boolean = showSetting

    override fun closeSetting() {
        showSetting = false
    }

    var user: FBUser? by mutableStateOf(null)
        private  set
    var roomList: List<RoomInfo> by mutableStateOf(emptyList())
        private  set
    var roomId: String by mutableStateOf("")
        private  set
    var totalRoomInfoValue: TotalRoomInfo by mutableStateOf(TotalRoomInfo.getInit())
        private set

    var syncData: Boolean by mutableStateOf(false)
        private  set
    var showLoading: Boolean by mutableStateOf(false)
        private  set
    var msg: String by mutableStateOf("")
        private  set
    var sheetSpace: Float by mutableFloatStateOf(20f)
        private  set
    var defaultUrl: String by mutableStateOf(ILOA)
        private  set
    init {
        viewModelScope.launch {
            userFlow.collect {
                user = it
            }
        }
        viewModelScope.launch {
            roomListFlow.collect {
                roomList = it
            }
        }
        viewModelScope.launch {
            roomIdFlow.collect {
                roomId = it
            }
        }
        viewModelScope.launch {
            totalRoomInfoFlow.collect {
                totalRoomInfoValue = it
            }
        }
        viewModelScope.launch {
            syncDataFlow.collect {
                syncData = it
            }
        }
        viewModelScope.launch {
            showLoadingFlow.collect {
                showLoading = it
            }
        }
        viewModelScope.launch {
            msgFlow.collect {
                msg = it
            }
        }
        viewModelScope.launch {
            sheetSpaceFlow.collect {
                sheetSpace = it
            }
        }
        viewModelScope.launch {
            defaultUrlFlow.collect {
                defaultUrl = it
            }
        }
    }
}

File: iosApp/iosApp/viewModel/LoaCellViewModel.swift
//
//  LoaCellViewModel.swift
//  iosApp
//
//  Created by WonHee Jung on 2023/07/06.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import FirebaseAuth
import shared
import SwiftUI
import SwiftUI_Snackbar
import Combine

class LoaCellViewModel: CommonViewModel, ObservableObject {
    
    let sc : SnackbarController = SnackbarController()

    @Published var roomList : [RoomInfo] = []
    @Published var user : FBUser? = nil
    @Published var roomId : String = ""
    @Published var totalRoomInfo : TotalRoomInfo = TotalRoomInfo.companion.getInit()
    @Published var showSplash :Bool = true
    @Published var defaultSpace : CGFloat = 20.0
    @Published var syncData : Bool = false
    @Published var showSetting :Bool = false
    override func closeSetting() {
        showSetting = false
    }
    
    override func getSetting() -> Bool {
        return showSetting
    }
    @Published var showLoading : Bool = false
    @Published var msg :String = ""
    @Published var baseUrl :String = ""
    
    
    var roomInfo :RoomInfo? {
        totalRoomInfo.roomInfo
    }
    var userInfo : shared.UserInfo? {
        totalRoomInfo.userInfo
    }
    var raidInfo : RaidInfo? {
        totalRoomInfo.raidInfo
    }
    var focusUserName : String {
        totalRoomInfo.focusUserName
    }
    var focusRaidId :String {
        totalRoomInfo.focusRaidId
    }
    var tabState : RoomState {
        totalRoomInfo.tabState
    }
    var dialogStatus : Modal? {
        totalRoomInfo.dialogState
    }
    var characterList : [Character] {
        totalRoomInfo.characterList
    }
    
    init() {
        super.init(
            dataBase: ModuleProvider().getAppDataBase(),
            config: ModuleProvider().getConfig(),
            loginHelper: ModuleProvider().getLoginHelper(),
            synergyReferenceHelper: ModuleProvider().getSynergyReferenceHelper()
        )

        loginHelper.auth.user.collect { user in
            self.user = user
            self.showSplash = false
        }
        syncDataFlow.collect { value in
            withAnimation {
                self.syncData = value!.boolValue
            }
        }
        roomListFlow.collect { value in
            self.roomList = value as! [RoomInfo]
        }
        roomIdFlow.collect { value in
            self.roomId = value! as String
        }
        totalRoomInfoFlow.collect { value in
            withAnimation {
                self.totalRoomInfo = value!
            }
        }
        showLoadingFlow.collect { value in
            withAnimation {
                self.showLoading = value as! Bool
            }
        }
        msgFlow.collect { value in
            self.msg = value! as String
        }
        sheetSpaceFlow.collect { value in
            self.defaultSpace = value as! CGFloat
        }
        defaultUrlFlow.collect { value in
            self.baseUrl = value! as String
        }
    }

    
    func fbUserIsAnonymous() -> KotlinBoolean? {
        guard let user = user else {
            return nil
        }
        return KotlinBoolean(bool: user.isAnonymous)
    }
    
    
    func getUserUid() -> String? {
        return user?.uid
    }
    
    
    func showRoomEnterError() {
        
    }
    
    override func showSnackBar(msg: String) {
        self.sc.showSnackBar(message: msg,label: "확인")
    }
    
    func showSnackBar(msg: String,label:String,action:@escaping () -> Void) {
        self.sc.showSnackBar(message: msg,label: label, perfromAction : action)
    }
    
    func resetSnackBar() {
        self.sc.resetSnackBar()
    }
    
}
